localstack-core/localstack/packages/java.py

```diff
@@ -0,0 +1,138 @@
+
+    def _get_download_url(self) -> str:
+        try:
+            LOG.debug("Determining the latest Java build version")
+            return self._download_url_latest_release()
+        except Exception as exc:  # noqa
+            LOG.debug(
+                "Unable to determine the latest Java build version. Using pinned versions: %s", exc
+            )
+            return self._download_url_fallback()
+
+    def _post_process(self, target: InstallTarget) -> None:
+        target_directory = self._get_install_dir(target)
+        minimal_jre_path = os.path.join(target.value, self.name, f"{self.version}.minimal")
+        rm_rf(minimal_jre_path)
+
+        # If jlink is not available, use the environment as is
+        if not os.path.exists(os.path.join(target_directory, "bin", "jlink")):
+            LOG.warning("Skipping JRE optimisation because jlink is not available")
+            return
+
+        # Build a custom JRE with only the necessary bits to minimise disk footprint
+        LOG.debug("Optimising JRE installation")
+        cmd = (
+            "bin/jlink --add-modules "
+            # Required modules
+            "java.base,java.desktop,java.instrument,java.management,"
+            "java.naming,java.scripting,java.sql,java.xml,jdk.compiler,"
+            # jdk.unsupported contains sun.misc.Unsafe which is required by some dependencies
+            "jdk.unsupported,"
+            # Additional cipher suites
+            "jdk.crypto.cryptoki,"
+            # Archive support
+            "jdk.zipfs,"
+            # Required by MQ broker
+            "jdk.httpserver,jdk.management,jdk.management.agent,"
+            # Required by Spark and Hadoop
+            "java.security.jgss,jdk.security.auth,"
+            # OpenSearch requires Thai locale for segmentation support
+            "jdk.localedata --include-locales en,th "
+            # Supplementary args
+            "--compress 2 --strip-debug --no-header-files --no-man-pages "
+            # Output directory
+            "--output " + minimal_jre_path
+        )
+        run(cmd, cwd=target_directory)
+
+        rm_rf(target_directory)
+        os.rename(minimal_jre_path, target_directory)
+
+    def get_java_home(self) -> str:
+        """
+        Get JAVA_HOME for this installation of Java.
+        """
+        return self.get_installed_dir()
+
+    @property
+    def arch(self) -> str:
+        return (
+            "x64" if get_arch() == Arch.amd64 else "aarch64" if get_arch() == Arch.arm64 else None
+        )
+
+    def _download_url_latest_release(self) -> str:
+        """
+        Return the download URL for latest stable JDK build.
+        """
+        endpoint = (
+            f"https://api.adoptium.net/v3/assets/latest/{self.version}/hotspot?"
+            f"os=linux&architecture={self.arch}&image_type=jdk"
+        )
+        response = requests.get(endpoint, headers={"user-agent": "example/0.0.0"}).json()
```

Comment:
q: The user agent setting here is a bit unorthodox. Is there a reason why you explicitly set it here? I think - if possible - you should either not set the user agent at all (fall back to `requests`), or set the LocalStack version as user agent.

Reply:
Yes, this could've been more elegant. For whatever reason, the Adoptium API denies service to the requests client

```
$ curl 'https://api.adoptium.net/v3/assets/latest/11/hotspot?os=linux&architecture=x64&image_type=jdk' -H user-agent:python-requests/2.32.3
<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
<hr><center>Microsoft-Azure-Application-Gateway/v2</center>
</body>
</html>

$ curl 'https://api.adoptium.net/v3/assets/latest/11/hotspot?os=linux&architecture=x64&image_type=jdk' -H user-agent:
[
    {
        "binary": {
            "architecture": "x64",
...
```

In 680a58a65, I define and use a standard LS user-agent string.